### PR TITLE
Alt. Router Initial Work

### PR DESCRIPTION
**Pull request by vsavkin**

### CODE REVIEW

File: modules/angular2/test/alt_router/router_url_parser_spec.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  it,
+  xit
+} from 'angular2/testing_internal';
+
+import {DefaultRouterUrlParser} from 'angular2/src/alt_router/router_url_parser';
+import {UrlSegment} from 'angular2/src/alt_router/segments';
+
+export function main() {
+  describe('url parsing', () => {
+    let parser = new DefaultRouterUrlParser();
```
Comment (vsavkin) [inline modules/angular2/test/alt_router/router_url_parser_spec.ts]: UrlMarshaller? UrlSerializer/UrlDeserializer?

File: modules/angular2/src/alt_router/segments.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  constructor(public current: T, public children: TreeNode<T>[]) {}
+}
+
+export class UrlSegment {
+  constructor(public segment: string, public parameters: {[key: string]: string},
+              public outlet: string) {}
+}
+
+export class RouteSegment {
+  constructor(public urlSegments: UrlSegment[], public parameters: {[key: string]: string},
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/segments.ts]: Add the router field

File: modules/angular2/src/alt_router/router.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+import {recognize} from './recognize';
+import {RouteParams} from './route_params';
+import {RouteSegment, TreeNode} from './segments';
+
+export abstract class Router {
+  abstract navigateByUrl(url: string): Promise<any>;
+  abstract registerOutlet(name: string, outlet: RouterOutlet): void;
+}
+
+export class RootRouter extends Router {
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/router.ts]: Rename into Router

File: modules/angular2/src/alt_router/router.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        .then(routeSegments => {
+          _loadSegments(routeSegments, this._prevSegments, this, this);
+          this._prevSegments = routeSegments;
+        });
+  }
+
+  registerOutlet(name: string, outlet: RouterOutlet): void { this._outlets[name] = outlet; }
+}
+
+export class ChildRouter extends Router {
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/router.ts]: Call it RouterOutletMap?

File: modules/angular2/src/alt_router/recognize.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
+import {RouteSegment, UrlSegment, TreeNode} from './segments';
+import {RoutesMetadata, RouteMetadata} from './metadata/metadata';
+import {Type, isPresent, stringify} from 'angular2/src/facade/lang';
+import {PromiseWrapper} from 'angular2/src/facade/promise';
+import {BaseException} from 'angular2/src/facade/exceptions';
+import {ComponentResolver} from 'angular2/core';
+import {reflector} from 'angular2/src/core/reflection/reflection';
+
+export function recognize(componentResolver: ComponentResolver, type: Type,
+                          url: TreeNode<UrlSegment>): Promise<TreeNode<RouteSegment>> {
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/recognize.ts]: Tree

File: modules/angular2/src/alt_router/router.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import {recognize} from './recognize';
+import {equalSegments, routeSegmentComponentFactory, RouteSegment, TreeNode} from './segments';
+
+export class RouterOutletMap {
+  /** @internal */
+  _outlets: {[name: string]: RouterOutlet} = {};
+  registerOutlet(name: string, outlet: RouterOutlet): void { this._outlets[name] = outlet; }
+}
+
+export class Router extends RouterOutletMap {
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/router.ts]: Make it contain an outlet map

File: modules/angular2/src/alt_router/segments.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    let index = this._nodes.indexOf(t);
+    return index > -1 && index < this._nodes.length - 1 ? [this._nodes[index + 1]] : [];
+  }
+
+  firstChild(t: T): T {
+    let index = this._nodes.indexOf(t);
+    return index > -1 && index < this._nodes.length - 1 ? this._nodes[index + 1] : null;
+  }
+
+  pathToRoot(t: T): T[] {
```
Comment (vsavkin) [inline modules/angular2/src/alt_router/segments.ts]: rename into pathFromRoot

**Comment by vsavkin** [comment]
Did a walkthrough with Misko

File: modules/angular2/src/alt_router/router_url_parser.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+export class DefaultRouterUrlParser extends RouterUrlParser {
+  parse(url: string): Tree<UrlSegment> {
+    if (url.length === 0) {
+      throw new BaseException(`Invalid url '${url}'`);
+    }
+    return new Tree<UrlSegment>(this._parseNodes(url));
+  }
+
+  private _parseNodes(url: string): UrlSegment[] {
```
Comment (gkalpak) [inline modules/angular2/src/alt_router/router_url_parser.ts]: Isn't this method essentially doing `url.split('/').map(segment => new UrlSegment(segment, {}, ''))` (with the extra twist that a leading `/` is part of the first segment) ?

File: modules/angular2/src/alt_router/segments.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import {ComponentFactory} from 'angular2/core';
+import {StringMapWrapper, ListWrapper} from 'angular2/src/facade/collection';
+import {Type, isBlank} from 'angular2/src/facade/lang';
+
+export class Tree<T> {
```
Comment (gkalpak) [inline modules/angular2/src/alt_router/segments.ts]: OOC, why is this called `Tree`. Are there any plans to allow multiple children in the future ?

**Comment by rkirov** [comment]
depends on tobias' changes so this is skipping g3_v2_0 branch

**Pull request closed without merging**